### PR TITLE
Enable viewing and managing filters

### PR DESCRIPTION
**Pull request by geraldaddey**
## Description

The implemented filter and filter tasks functionality allows users to:

- View filters from the home and quicklink views
- View filter tasks 
- Add/Remove filters from favorites. 

Potential Feature Implmentations in the Future
- Copy Filter Query: Enable users to copy and share filter queries.
- Group & Sort by Filters: Allow users to group filtered data based on specific criteria.
- Remove from Filter: Provide an option to remove tasks from a filter.

## Screencast
screenshot: 234
screenshot: 231
screenshot: 235
screenshot: 236
screenshot: 237
screenshot: 238

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

File: extensions/todoist/package.json [hunk 1]
```diff
@@ -204,6 +204,12 @@
       "description": "Show your Todoist labels.",
       "mode": "view"
     },
+    {
+      "name": "show-filters",
+      "title": "Show Filters",
+      "description": "Show filters based on given query",
```
Comment (thomaslombart) [inline extensions/todoist/package.json:210]:
```suggestion
      "description": "Show your Todoist filters.",
```

File: extensions/todoist/src/api.ts [hunk 2]
```diff
@@ -568,6 +570,68 @@ export async function deleteLabel(id: string, { data, setData }: CachedDataParam
...
+        uuid: crypto.randomUUID(),
+        args,
+      },
+    ],
+  });
+
+  if (data) {
+    setData({
+      ...data,
+      labels: data.labels.map((p) => (p.id === args.id ? updatedData.labels[0] : p)),
```
Comment (thomaslombart) [inline extensions/todoist/src/api.ts:607]: It should probably be filters here.

File: extensions/todoist/src/api.ts [hunk 3]
```diff
@@ -568,6 +570,68 @@ export async function deleteLabel(id: string, { data, setData }: CachedDataParam
...
+        uuid: crypto.randomUUID(),
+        args: { id },
+      },
+    ],
+  });
+
+  if (data) {
+    setData({
+      ...data,
+      labels: data.labels.filter((l) => {
```
Comment (thomaslombart) [inline extensions/todoist/src/api.ts:628]: Same thing here.

File: extensions/todoist/src/components/FilterTasks.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        .then((fetchedTasks) => {
+          setTasks(filterSort(fetchedTasks));
+        })
+        .catch(() => {
+          throw new Error("Error fetching filter tasks");
+        });
+    } else {
+      setTasks([]);
+    }
+  }, [query]);
```
Comment (thomaslombart) [inline extensions/todoist/src/components/FilterTasks.tsx:34]: Why not using `useCachedPromise`? It'll make the code a bit more readable and you can benefit from caching.

File: extensions/todoist/src/show-filters.tsx [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+        await showToast({ style: Toast.Style.Success, title: "Filter deleted" });
+      } catch (error) {
+        handleError({ error, title: "Unable to delete filter" });
+      }
+    }
+  }
+
+  return (
+    <List searchBarPlaceholder="Filter task filters by name">
```
Comment (thomaslombart) [inline extensions/todoist/src/show-filters.tsx:61]:
```suggestion
    <List searchBarPlaceholder="Filter filters by name">
```

**Review by thomaslombart — CHANGES_REQUESTED** [review]
Awesome additions! 💪

Could you add an entry in the `CHANGELOG.md` file? Here's our guidelines about that: [Version History](https://developers.raycast.com/basics/prepare-an-extension-for-store#version-history).

**Comment by geraldaddey** [comment]
Thanks for the review @thomaslombart. I will work on the requested changes 🫡

**Comment by thomaslombart** [comment]
Hey @geraldaddey, is the PR ready to review again?

File: extensions/todoist/src/components/FilterTasks.tsx (hunk 4 above)
Comment (geraldaddey) [reply extensions/todoist/src/components/FilterTasks.tsx:34]: thanks for the tip. changed the implementation to make use of useCachedPromise.

**Comment by geraldaddey** [comment]
Hi @thomaslombart the changes requested have been effected. Please you can review again and let me know if anything comes up

**Comment by geraldaddey** [comment]
after implementing my filterTasks using the useCachedPromise, i am experiencing issues using the tasks in the useViewTasks because it expects data  from syncData but fetching these filter tasks is only available in the Rest API. can you help here so i can wrap up and have a good PR

**Comment by thomaslombart** [comment]
Hey @geraldaddey, I don't have much bandwidth for that right now as I'm focused on other projects. I'd suggest reaching out to the community if you still have these issues and want to quickly wrap it up 😊

**Comment by geraldaddey** [comment]
@thomaslombart please review again. i have made the changes and now using usedCachedPromise for the filter tasks

File: extensions/todoist/src/components/FilterTasks.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    return filterTasks;
+  };
+
+  const { isLoading, data, revalidate } = useCachedPromise(getFilterTasksCached, [query]);
+
+  useEffect(() => {
+    if (query) {
+      revalidate();
+    }
+  }, [query]);
```
Comment (thomaslombart) [inline extensions/todoist/src/components/FilterTasks.tsx:34]: This doesn't seem to be needed as you've already past `query` as a dependency in `useCachedPromise`.

File: extensions/todoist/src/components/FilterTasks.tsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    );
+  }
+
+  return (
+    <TaskListSections
+      mode={ViewMode.project}
+      sections={viewProps.groupBy?.value === "default" ? [{ name, tasks: tasks }] : sections}
+      viewProps={viewProps}
+      quickLinkView={quickLinkView}
+    />
```
Comment (thomaslombart) [inline extensions/todoist/src/components/FilterTasks.tsx:71]: You can also pass the `isLoading` boolean over here.

File: extensions/todoist/src/components/FilterTasks.tsx (hunk 6 above)
Comment (geraldaddey) [reply extensions/todoist/src/components/FilterTasks.tsx:34]: @thomaslombart  thanks for pointing that out .

File: extensions/todoist/src/components/FilterTasks.tsx (hunk 7 above)
Comment (geraldaddey) [reply extensions/todoist/src/components/FilterTasks.tsx]: done. thanks for the tips. Would definitely explore and fix some more issues

**Review by thomaslombart — APPROVED** [review]
🚀

**Comment by erauner12** [comment]
Hey guys, I noticed that any filters that contain commas are broken. Should I go ahead and log an issue for this bug?

**Comment by thomaslombart** [comment]
Sure, feel free to do so!

**Pull request merged**